### ISSUE TITLE
Revamping stage plugins (Removing the stage-x presets)

### ISSUE THREAD

**Issue by thejameskyle** [babel/babel#4955]
So there's a problem with how stage-x plugins are designed right now. They are versioned in lockstep with the rest of Babel. However these stage-x features change throughout the process and may make breaking changes at any time.

Babel's lockstep versioning causes a weird tradeoff of either following SemVer or keeping up with the spec. At the same time Lockstep versioning is important for the ecosystem.

I've talked to contributors privately about this and figured I'd create an issue:

I want to get rid of the `babel-preset-stage-x` plugins and move all of the stage-x plugins into their own monorepo with independent (see Lerna) versioning.

This way there are no more presets for stages and people have to manually enable plugins by individual features. These plugins can then make major version bumps at any time.

The plugins should be named `babel-plugin-transform-proposal-{feature}`, when they are accepted by TC39 they move into the main repository and are named `babel-plugin-transform-es{year}-{feature}` and included in the relevant presets.

The one problem that needs to be solved is parsing. We need to be able to change the syntax of a proposal at any time so that we can track changes like decorators moving from `@` to `#` (which was turned down by the committee but we need a better way in Babel to handle that type of scenario).

Babylon has the ability to write syntax features as plugins (see https://github.com/babel/babylon/tree/master/src/plugins), but right now es proposals are embedded within the parser. Instead we should build es proposal syntax as plugins.

Right now Babylon does not allow external plugins to be added in order to prevent the ecosystem from going mad with power using them.

I would like to suggest adding a whitelist to Babylon for allowed parser plugins in Babylon and allowing them to be included in Babel plugins like:

```js
export default function babelPlugin() {
  return {
    __internal_only_dont_use_this_or_we_will__find_you_and_intentionally_break_your_plugin_in_a_patch_release__you_have_been_warned__: {
      __dangerous_dont_use__: {
        __seriously_no__: {
          __just_fork_babylon_yourself__: {
            parserPlugin: { name: 'proposal-{feature}', version: require('./package.json').version, babelVersions: '^6.0.0', plugin: require('./parser-plugin') }
          }
        }
      }
    }
  };
}
```

I think that should be enough.

It's gonna be controversial to remove the stage presets, but it is the right thing to do for the ecosystem in my opinion.

**Comment by hzoo** [comment]
@thejameskyle feel free to add this issue to the blog post as a "relevant issue"

**Comment by thejameskyle** [comment]
I'd rather hear other contributors thoughts before we invite the entire community to come comment on this, that can come later.

**Comment by hzoo** [comment]
> The plugins should be named babel-plugin-transform-proposal-{feature}, when they are accepted by TC39 they move into the main repository and are named babel-plugin-transform-es{year}-{feature} and included in the relevant presets.

Sounds good - would it be useful to include the stage of the proposal in the package name and then just make a new package?

`babel-plugin-transform-proposal-stage-0-{feature}`
`babel-plugin-transform-proposal-stage-1-{feature}`

or just bump majors? Putting it in the name could be nice as well. Would it be crazy that we can even add the month of the meeting that the spec changed instead? Basically no major versions and just make a new package

**Comment by thejameskyle** [comment]
Is it just the name that would be the benefit? Major version bumps seem to cover everything else fine

**Comment by hzoo** [comment]
Yeah just the name. Majors could be more useful and maybe we can add that where 0.x is stage-0 and 1.x is stage-1? not sure

**Comment by thejameskyle** [comment]
I wouldn't want the versioning to be tied to the stage, proposals can change within a single stage.

**Comment by hzoo** [comment]
👍

**Comment by chicoxyzzy** [comment]
It would be great to move proposal plugins to their own repo. So we'll never have something like `babel-plugin-transform-decorators-legacy` again. But what about relation to version of Babel? If we'll increase major version of plugin when Babel will update to version 7, we'll loose an ability to maintain plugin version for Babel 6. This is a real problem IMO.

**Comment by hzoo** [comment]
> If we'll increase major version of plugin when Babel will update to version 7, we'll loose an ability to maintain plugin version for Babel 6. This is a real problem IMO.

Those plugins should be versioned independently to Babel? What do you mean here?

**Comment by chicoxyzzy** [comment]
Let's say we `babel-plugin-transform-proposal-awesome-feature` version 3.0.0.
Than following semver:
1. we found a bug and fixed it - v3.0.1
2. we add some options (loose mode, etc.) - v3.1.0
3. proposal spec changes and we update plugin - v4.0.0

Now we release Babel 7 with breaking changes for plugins.
What should we do now?

We can bump major version of `babel-plugin-transform-proposal-awesome-feature` to v5.0.0, but what if spec changes another time after that? Now we should bump both v4.0.0 and v5.0.0 to support latest spec, but we can't

**Comment by chicoxyzzy** [comment]
The first solution I see here is to stick major version of plugin to Babel version. But then we should use minor for breaking changes and patch for the rest.

The second solution is to add Babel version to plugin name (`babel6-plugin-transform-proposal-awesome-feature`)

**Comment by chicoxyzzy** [comment]
Same problem for ES features.

**Comment by taion** [comment]
It might make sense to adopt different versioning strategies by stage to better convey relative levels of maturity. Also, for the rougher proposal, dates (of the relevant TC39 meeting) are a natural versioning scheme – they have a definite meaning, and SemVer allows bumping versions by more than 1 anyway.

Here's one way you could do it:
- For stage 0, use only the patch version: `v0.0.20161215-stage0.2`
- For stage 1, use only the minor version: `v0.20161215.1`
- For stage 2+, use standard SemVer (starting at v1.0.0)

Other variations are possible, but I think using dates in the version string makes a lot of sense for stage 0 and stage 1 proposals, and if you want to indicate that something is not in a good state for production use, then prereleases express that idea quite well as well.

**Comment by hzoo** [comment]
That sounds pretty interesting as well. We can combine that with what james was saying:

> The plugins should be named babel-plugin-transform-proposal-{feature}, when they are accepted by TC39 they move into the main repository and are named babel-plugin-transform-es{year}-{feature} and included in the relevant presets.

Only issue is solving the parser issue - transform Babel plugins will need to specify which parser plugins they support to make them independently versioned but not sure how that conflicts with one another

**Comment by hzoo** [comment]
> But than we should use minor for breaking changes (according to proposal spec) and patch for the rest (like now). This is ComVer-ish solution.

ComVer is supposed to be compatible with SemVer - it's just major = breaking, minor = anything else so not what you described?

**Comment by chicoxyzzy** [comment]
Right. What I'm trying to say that this way there is no difference between bug fix (patch) and feature add (minor)

**Comment by taion** [comment]
One issue with `babel-plugin-transform-proposal-{feature}` is that "proposal" is potentially ambiguous with "stage 1". Name collisions 😞

**Comment by bfred-it** [comment]
_I haven't read every comment above._

I think semver already covers the situation fully. Let each transform have its own independent version.

1. As a developer using, let's say, `function-bind`, I'll install `babel-plugin-transform-function-bind@1.0.0`
2. Tomorrow you fix something: _1.0.1 is pushed out_
3. The proposal changes slightly or you implement a new part of it: _1.1.0_
4. The proposal changes entirely, `::` becomes `$%^`: _here comes 2.0.0_
5. The proposal steps through stages 1, 2, 3, 4 and lands in ES2023: _you do absolutely nothing_

Pros:

- My install will never break because ~1.0.0 will stop at 1.1.0 forever
- You will never need any packages like like `*-legacy` because you can already `npm install transform-xxx@1.1.0`
- You won't need to update anything when a proposal steps through each stage.

Cons:

- I don't know how this versioning fits your Lerna workflow, but it may be worth looking into.

**Comment by bfred-it** [comment]
For presets, do the same, but of course they also need major updates every time any dependency has a major update. I think that's the expectation with semver, yes?

And yes, stage-0 and stage-1 should probably go.

**Comment by hzoo** [comment]
> I don't know how this versioning fits your Lerna workflow, but it may be worth looking into.

This is the reasoning behind moving anything experimental into a separate repo. However babel-types/babel-generator are currently their own modules and are not pluggable atm either. And the real issue isn't about the transform plugins but the parser. If you update the syntax and we need to change the parser/babylon then we need to make a major version bump, or create a versioning system in babylon itself (no), or make the syntax plugins independent modules as well.

(You should probably read the comments above since it does go into this issue). Also the reason for the naming is not about semver/versioning but about getting the idea across that these are experimental/proposals rather than everyone using stage-0 in their prod apps just because "less config" or it's copied from another project.

**Comment by bfred-it** [comment]
Not sure I get what you're saying; is that not solved by having `transform-xxx@2` have a minimum supported version of babylon (or parser plugins)?

> copied from another project.

Any naming you put in there won't change the fact that people will copy anything blindly. Following semver for each transform will at least save them from breaking changes. Probably doing away with stage-0 and stage-1 will make it less easy to make the mistake for anyone else.

**Comment by Daniel15** [comment]
> their own monorepo with independent (see Lerna) versioning.

How would we version [babel-standalone](https://github.com/babel/babel-standalone) in that case? Currently it's easy as the version of babel-standalone is just max(version of all Babel packages)

**Comment by thejameskyle** [comment]
@Daniel15 Treat these plugins like you treat babili in the browser.
